### DISCUSSION TITLE
Re-writing this SKBitmap code with C# Pointers rather than using SKBitmap.Pixels?

### DISCUSSION THREAD

**Discussion by jonmdev** [mono/SkiaSharp#2626] [Q&A]
I am using the following SkiaSharp code to apply a multiply effect (tint) from a given `tintColor` to the image and create a new SKBitmap with the applied tint:

```
                SKColor[] pixels = skBitmap.Pixels;
                for (int i=0; i < pixels.Length; i++) {
                    pixels[i] = new SKColor((byte)(pixels[i].Red * tintColor.Red), (byte)(pixels[i].Green * tintColor.Green), (byte)(pixels[i].Blue * tintColor.Blue), pixels[i].Alpha);
                }
                tintedBitmap = new SKBitmap(skBitmap.Width, skBitmap.Height);
                tintedBitmap.Pixels = pixels;
```

This is working fine, but it seems evident to me from the documentation [here][1] that there is some inefficiency.

In particular, the line `SKColor[] pixels = skBitmap.Pixels` copies the entire pixel data over into this array which is not necessary. I could alternatively create an empty array like `SKColor[] pixels = new SKColor[skBitmap.Width * skBitmap.Height]` and then fill this in the for loop by accessing the `skBitmap` color data with pointers. 

However, I am having a hard time understanding how to use pointers to do this. I read the documentation linked above but the C# pointers and byte management of the colors is a bit above my head.

What might the for loop in this example look like if using pointers to reference the color data from `skBitmap` instead?

Thanks for any help.

  [1]: https://learn.microsoft.com/en-us/xamarin/xamarin-forms/user-interface/graphics/skiasharp/bitmaps/pixel-bits

**Comment by mattleibow** [comment]
You can try using the pixel data directly.

This is much easier through the `SKPixmap` type, and you can get that with `skBitmap.PeekPixels()` and then with `SKPixmap.GetPixelSpan<SKColor>()`.

One thing to not is that you will need to verify if this is RGB or BGR pixel ordering. You can do this by checking the color type.
